nx/lib/nx/shape.ex

```diff
@@ -1218,6 +1218,57 @@ defmodule Nx.Shape do
     {List.to_tuple(result_shape), result_names}
   end
 
+  @doc """
+  Returns shape if valid and raises error if not.
+  """
+  def take_diagonal(tensor) do
+    case tensor.shape do
+      {len, breadth} ->
+        {len, breadth}
+
+      _bad_shape ->
+        raise ArgumentError, "take_diagonal/2 expects tensor of rank 2, got:\n#{inspect(tensor)}"
+    end
+  end
+
+  @doc """
+  Validates an offset for given shape
+
+  ## Examples
+
+      iex> Nx.Shape.validate_offset!({3, 4}, 1)
+      :ok
+
+      iex> Nx.Shape.validate_offset!({3, 4}, -1)
+      :ok
+
+  ## Error cases
+
+    iex> Nx.Shape.validate_offset!({3, 4}, 3)
+    ** (ArgumentError) offset must be less than length when positive, got: 3
+
+    iex> Nx.Shape.validate_offset!({3, 4}, -4)
+    ** (ArgumentError) absolute value of offset must be less than breadth when negative, got: -4
+  """
+  def validate_offset!(shape, offset) do
+    {len, breadth} = shape
+
+    case offset do
+      i when i >= 0 and i < len ->
+        :ok
+
+      i when i >= 0 ->
+        raise ArgumentError, "offset must be less than length when positive, got: #{inspect(i)}"
+
+      i when i < 0 and -i < breadth ->
+        :ok
+
+      i when i < 0 ->
+        raise ArgumentError,
+              "absolute value of offset must be less than breadth when negative, got: #{inspect(i)}"
+    end
```

Comment:
Shouldn't this be either a cond statement or a series of function clauses with guards instead?

Reply:
At first thought, function clauses and guards seem more appropriate. However, when we generalise, `cond` would be more appropriate.

WDYT?

Reply:
Both would be fine to be. It's just that the case statement adds a new variable without need

Reply:
@polvalente, 

On second thoughts, is the extra variable really a problem? Using `case` allows me to pattern match within a function. If I use `cond`, I will have to give up the readability of pattern matching, and if I use function clauses, I will have to give up the readability of having everything in one function.

Reply:
>  If I use cond, I will have to give up the readability of pattern matching

The issue here is that you are not really pattern matching on anything. The `i when` is redundant on all clauses and that's exactly what `cond` will help address. :)